src/foam/nanos/crunch/JavaCrunchService.java

```diff
@@ -0,0 +1,140 @@
+package foam.nanos.crunch;
+
+import foam.core.X;
+import foam.dao.ArraySink;
+import foam.dao.DAO;
+import foam.mlang.predicate.Predicate;
+import foam.nanos.auth.Subject;
+import foam.nanos.auth.User;
+import foam.nanos.logger.Logger;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import static foam.mlang.MLang.*;
+
+public class JavaCrunchService implements CrunchService {
+  public List getGrantPath(X x, String rootId) {
+    Logger logger = (Logger) x.get("logger");
+
+    DAO prerequisiteDAO = (DAO) x.get("prerequisiteCapabilityJunctionDAO");
+    DAO capabilityDAO = (DAO) x.get("capabilityDAO");
+
+    // Lookup for indices of previously observed capabilities
+    Map<String,Integer> alreadyListed = new HashMap<String,Integer>();
+    // Return list (list is reversed at the end to put prerequisites first)
```

Comment:
I'm planning on making a `.flow` document to explain how to use CRUNCH, maybe that would be best suited there. The prerequisites are returned first because `getGrantPath` returns capabilities in the same order which a user is expected to acquire them.

Reply:
Your comment is explaining what the code is, not why. Some one reviewing your code could see what this comment describes by looking at the code. It's helpful to have documentation explaining why the code is doing what it is doing.

Reply:
Oh I misunderstood what you meant. I figured that comment would be helpful as a warning since the reversal is easy to miss. I'll update it.